packages/repository/src/define-repository-class.ts

```diff
@@ -34,62 +34,73 @@ export interface CrudRepositoryClass<
-
 /**
  * Signature for repository classes that can be used as the base class for
- * `define*` functions
+ * `define*` functions. The constructor of a base repository class accepts
+ * the target model constructor and the datasource to use.
+ *
+ * `define*` functions require a class implementing this interface on input.
+ *
+ * @typeParam M - Model class constructor, e.g `typeof Model`.
+ * **❗️IMPORTANT: The type argument `M` is describing the model constructor type
+ * (e.g. `typeof Model`), not the model instance type (`Model`) as is the case
+ * in other repository-related types. The constructor type is required
+ * to support custom repository classes requiring a Model subclass in the
+ * constructor arguments, e.g. `Entity` or a user-provided model.**
  *
- * @typeParam M - Model class
  * @typeParam R - Repository class/interface
  */
-export interface BaseRepositoryClass<M extends Model, R extends Repository<M>> {
+export interface BaseRepositoryClass<
+  M extends typeof Model,
+  R extends Repository<PrototypeOf<M>>
+> {
   /**
    * The constructor for the generated repository class
    * @param modelClass - Model class
    * @param dataSource - DataSource object
    */
-  new (
-    // Make model class conditional of Model or Entity
-    modelClass: (M extends Entity ? typeof Entity : typeof Model) & {
-      prototype: M;
-    },
-    dataSource: juggler.DataSource,
-  ): R;
+  new (modelClass: M, dataSource: juggler.DataSource): R;
   prototype: R;
 }
+
 /**
- * Create (define) a CRUD repository class for the given model.
+ * Create (define) a repository class for the given model.
+ *
+ * See also `defineEntityCrudRepositoryClass` and `defineKeyValueRepositoryClass`
+ * for convenience wrappers providing repository class factory for the default
+ * CRUD and KeyValue implementations.
+ *
+ * **❗️IMPORTANT: The compiler (TypeScript 3.8) is not able to correctly infer
+ * generic arguments `M` and `R` from the class constructors provided in
+ * function arguments. You must always provide both M and R types explicitly.**
  *
  * @example
  *
  * ```ts
- * const AddressRepository = defineCrudRepositoryClass(Address);
+ * const AddressRepository = defineRepositoryClass<
+ *   typeof Address,
+ *   DefaultEntityCrudRepository<
+ *    Address,
+ *    typeof Address.prototype.id,
+ *    AddressRelations
+ *   >,
+ * >(Address, DefaultCrudRepository);
  * ```
  *
  * @param modelClass - A model class such as `Address`.
+ * @param baseRepositoryClass - Repository implementation to use as the base,
+ * e.g. `DefaultCrudRepository`.
  *
- * @typeParam M - Model class
- * @typeParam R - CRUD Repository class/interface
+ * @typeParam M - Model class constructor (e.g. `typeof Address`)
+ * @typeParam R - Repository class (e.g. `DefaultCrudRepository<Address, number>`)
  */
-export function defineCrudRepositoryClass<
-  M extends Model,
-  R extends CrudRepository<M>
+export function defineRepositoryClass<
+  M extends typeof Model,
```

Comment:
Similarly here, we need `M` to represent the class constructor, not the class instance.